Source/Scene/Cesium3DTileBatchTable.js

```diff
@@ -116,6 +131,121 @@ define([
+        var parentCounts = json.parentCounts;
+        var parentIds = json.parentIds;
+        var parentIdsLength = instancesLength;
+
+        if (defined(classIds.byteOffset)) {
+            classIds.componentType = defaultValue(classIds.componentType, 'UNSIGNED_SHORT');
+            classIds.type = 'SCALAR';
+            binaryAccessor = getBinaryAccessor(classIds);
+            classIds = binaryAccessor.createArrayBufferView(binary.buffer, binary.byteOffset + classIds.byteOffset, instancesLength);
+        }
+
+        var parentIndexes;
+        if (defined(parentCounts)) {
+            if (defined(parentCounts.byteOffset)) {
+                parentCounts.componentType = defaultValue(parentCounts.componentType, 'UNSIGNED_SHORT');
+                parentCounts.type = 'SCALAR';
+                binaryAccessor = getBinaryAccessor(parentCounts);
+                parentCounts = binaryAccessor.createArrayBufferView(binary.buffer, binary.byteOffset + parentCounts.byteOffset, instancesLength);
+            }
+            parentIndexes = new Uint16Array(instancesLength);
+            parentIdsLength = 0;
+            for (i = 0; i < instancesLength; ++i) {
+                parentIndexes[i] = parentIdsLength;
+                parentIdsLength += parentCounts[i];
+            }
+        }
+
+        if (defined(parentIds.byteOffset)) {
+            parentIds.componentType = defaultValue(parentIds.componentType, 'UNSIGNED_SHORT');
+            parentIds.type = 'SCALAR';
+            binaryAccessor = getBinaryAccessor(parentIds);
+            parentIds = binaryAccessor.createArrayBufferView(binary.buffer, binary.byteOffset + parentIds.byteOffset, parentIdsLength);
+        }
+
+        var classesLength = classes.length;
+        for (i = 0; i < classesLength; ++i) {
+            var classInstancesLength = classes[i].length;
+            var properties = classes[i].instances;
+            var binaryProperties = Cesium3DTileBatchTable.getBinaryProperties(classInstancesLength, properties, binary);
+            classes[i].instances = combine(binaryProperties, properties);
+        }
+
+        var classCounts = arrayFill(new Array(classesLength), 0);
+        var classIndexes = new Uint16Array(instancesLength);
+        for (i = 0; i < instancesLength; ++i) {
+            classId = classIds[i];
+            classIndexes[i] = classCounts[classId];
+            ++classCounts[classId];
+        }
+
+        var hierarchy = {
+            classes : classes,
+            classIds : classIds,
+            classIndexes : classIndexes,
+            parentCounts : parentCounts,
+            parentIndexes : parentIndexes,
+            parentIds : parentIds
+        };
+
+        //>>includeStart('debug', pragmas.debug);
+        validateHierarchy(hierarchy);
+        //>>includeEnd('debug');
+
+        return hierarchy;
+    }
+
+    //>>includeStart('debug', pragmas.debug);
+    function validateHierarchy(hierarchy) {
+        var stack = scratchStack;
```

Comment:
I would caution against using `scratchStack` here since it will have different debug/release usage and could lead to subtle bugs when new code is written.  To make the validation code self-contained just use a different scratch for it.